src/drivers/imx/dummy-dma.c

```diff
@@ -45,54 +163,361 @@ static int dummy_dma_status(struct dma_chan_data *channel,
+/**
+ * \brief Set channel configuration
+ * \param[in] channel: The channel to configure
+ * \param[in] config: Configuration data
+ * \return 0 on success, -EINVAL if the config is invalid or unsupported.
+ *
+ * Sets the channel configuration. For this particular driver the config means
+ * the direction and the actual SG elems for copying.
+ */
 static int dummy_dma_set_config(struct dma_chan_data *channel,
 				struct dma_sg_config *config)
 {
-	return 0;
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+	uint32_t flags;
+	int ret = 0;
+
+	spin_lock_irq(channel->dma->lock, flags);
+
+	if (!config->elem_array.count) {
+		trace_dummydma_error("dummy-dmac: %d channel %d no DMA descriptors",
+				     channel->dma->plat_data.id,
+				     channel->index);
+
+		ret = -EINVAL;
+		goto out;
+	}
+
+	channel->direction = config->direction;
+
+	if (config->direction != DMA_DIR_HMEM_TO_LMEM &&
+	    config->direction != DMA_DIR_LMEM_TO_HMEM) {
+		/* Shouldn't even happen though */
+		trace_dummydma_error("dummy-dmac: %d channel %d invalid direction %d",
+				     channel->dma->plat_data.id, channel->index,
+				     config->direction);
+		ret = -EINVAL;
+		goto out;
+	}
+	channel->desc_count = config->elem_array.count;
+	ch->elems = &config->elem_array;
+	ch->sg_elem_curr_idx = 0;
+
+	channel->status = COMP_STATE_PREPARE;
+out:
+	spin_unlock_irq(channel->dma->lock, flags);
+	return ret;
 }
 
 /* restore DMA context after leaving D3 */
 static int dummy_dma_pm_context_restore(struct dma *dma)
 {
+	/* Virtual device, no hardware registers */
 	return 0;
 }
 
 /* store DMA context after leaving D3 */
 static int dummy_dma_pm_context_store(struct dma *dma)
 {
+	/* Virtual device, no hardware registers */
 	return 0;
 }
 
 static int dummy_dma_set_cb(struct dma_chan_data *channel, int type,
 		void (*cb)(void *data, uint32_t type, struct dma_cb_data *next),
 		void *data)
 {
+	channel->cb = cb;
+	channel->cb_data = data;
+	channel->cb_type = type;
 	return 0;
 }
 
+/**
+ * \brief Get the next SG elem from the active configuration
+ * \param[in] channel: DMA channel to get the config from
+ * \param[out] next: Next element
+ * Return: true if @next was modified, false if no next element available.
+ *
+ * Get the next element from the config, so that the copy flow can continue.
+ */
+static bool get_next_elem(struct dma_chan_data *channel,
+			  struct dma_cb_data *next)
+{
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+	int curr_idx = ch->sg_elem_curr_idx;
+
+	if (curr_idx >= channel->desc_count)
+		return false;
+
+	ch->sg_elem_curr_idx++;
+
+	next->elem = ch->elems->elems[curr_idx];
+	return true;
+}
+
+/**
+ * \brief Perform one individual copy
+ * \param[in] ch: DMA channel to do the copy for
+ * \param[in, out] next: Next element
+ * \param[out] size: Size in bytes of the last completed transfer
+ * \return: 1 if transfer should stop immediately, 0 if it should continue.
+ *          Negative value on error (never happens)
```

Comment:
I don't like this naming convention with 1 and 0. We have two options:
* create meaningful return codes
* OR use boolean for return type. true -> if transfer should continue, false if trasnfer should stop immediately. We never really return an error :). 

I am in favor of the second option.